tb_plugin/torch_tb_profiler/profiler/overall_parser.py

```diff
@@ -177,22 +177,119 @@ def __init__(self):
+            # This step doesn't launch any device side event, just assign it as empty.
+            steps_device[i_step] = (step_device_min_ts, step_device_max_ts)
+            step_device_min_ts = sys.maxsize
+            step_device_max_ts = -sys.maxsize - 1
+            i_step += 1
+        # Change step time to device side on the condition that any step have device time.
+        is_use_gpu = (len(matched_device_nodes) > 0)
+        if is_use_gpu:
+            prev_step_end_time = self.steps[0][0]
+            if steps_device[0][0] != sys.maxsize:  # When step 0 has device event.
+                for device_node in device_node_list:
+                    if device_node not in matched_device_nodes:
+                        # Now this device_node is not launched inside any step span.
+                        if device_node.end_time < steps_device[0][0]:
+                            prev_step_end_time = max(prev_step_end_time, device_node.end_time)
+            for i_step in range(len(self.steps)):
+                step_start_time = max(prev_step_end_time, self.steps[i_step][0])
+                step_end_time = self.steps[i_step][1]
+                if steps_device[i_step][0] == sys.maxsize:  # When step i_step has no device event.
+                    # Assign to step_start_time when kernel is behind host step end.
+                    step_end_time = max(step_end_time, step_start_time)
+                else:
+                    step_end_time = max(step_end_time, steps_device[i_step][1])
+                    if step_end_time < step_start_time:
+                        logger.warning(
+                            "Abnormal step_end_time of step {}: [{}, {}]".format(
+                                i_step, step_start_time, step_end_time))
+                        step_end_time = step_start_time
+                self.steps[i_step] = (step_start_time, step_end_time)  # Update step time considering device side.
+                prev_step_end_time = step_end_time
+
+        is_remove_tail_steps = True  # TODO: Use tensorboard argument instead.
+        if is_use_gpu and len(self.steps) > 1 and is_remove_tail_steps:
+            i_step = len(self.steps) - 1
+            while i_step >= 0:
+                if steps_matched_device_nodes[i_step] > 0:
+                    break
+                i_step -= 1
+            if i_step >= 0:
+                keep_steps = i_step + 1
+                if i_step > 0 and steps_matched_device_nodes[i_step - 1] * 0.8 > steps_matched_device_nodes[i_step]:
+                    keep_steps = i_step
+                if keep_steps < len(self.steps):
+                    logger.warning(
+                        "Remove the last {} steps because of suspected lost of kernels in tracing file".format(
+                            len(self.steps) - keep_steps
+                        ))
+                    self.steps = self.steps[:keep_steps]
+                    self.steps_names = self.steps_names[:keep_steps]
```

Comment:
it is possible that some normal step may contain less kernels, e.g. gradient accumulation scenario.

as we discussed, find the last kernel, map it to step, then make exclusion decision.